server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala

```diff
@@ -443,8 +448,11 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
       }
     }
     if (null != diskOperators) {
+      cleanupExpiredShuffleKey(shuffleKeySet())
+      Thread.currentThread().join(RssConf.waitCleanTaskSubmitBeforeCloseTimeoutMs(conf))
```

Comment:
I guess the reason to wait here is for the clean task be submitted to diskOperator, but even we wait 500ms here, we are not sure the needed clean task are submitted. I think a better way is to delete the first executor.shutdown() in ThreadUtils.shutdown.

Reply:
Yes it is. But if we delete the `executor.shutdown()`, will it affect the logic of other places which call the `ThreadUtils.shutdown`, or other code  in the future  if they wanna call a graceful shutdown? I checked them once and feel it may not have effect on the current code, but for future code it is hard to say. Maybe we can tolerate some remaining dir & file if there is a huge size shuffle data that cannot be cleaned up in time,  then clean them when start in the place you mentioned before.

Reply:
FYI: In previous test, if not call thread.join before and just call ThreadUtil.shutdown, It can submit and process deletion tasks with a total amount of about almost 4~500GB file on SSD for one diskOperator (where is the cap didn't tested) before ThreadUtil.shutdown call the inner executor.shutdown.

Reply:
I may have some misunderstand in the previous comment. Let's discuss the process of your code.
1. The close thread calls cleanupExpiredShuffleKey, and all tasks to clean all data-file are submitted to diskOperator threadpool, to execute at some time in the future.
```
    threadPool.submit(new Runnable {
      override def run(): Unit = {
        deleteFileWithRetry(dir)
      }
    })
```
2. The ThreadPoolExecutors in diskOperators starts to run delete file tasks, at the same time the close thread starts waiting.
3. After 500ms, the close thread calls ThreadUtils.shutdown for diskOperators. ThreadUtils.shutdown calls executor.shutdown(), executor.awaitTermination() and executor.shutdownNow() in sequence.

So, diskOperators will run 500ms before executor.shutdown() is called. And if we do not call thread.join before ThreadUtil.shutdown, then executor.shutdown will be called immediately, so I guess the effect of calling thread.join is to let threadpool run 500ms before executor.shutdown() is called.

Maybe we can add a parameter called delay in ThreadUtils.shutdown, and sleep delay before calling executor.shutdown(), executor.awaitTermination() and executor.shutdownNow().

> I guess the reason to wait here is for the clean task be submitted to diskOperator, but even we wait 500ms here, we are not sure the needed clean task are submitted. I think a better way is to delete the first executor.shutdown() in ThreadUtils.shutdown.

Reply:
Oh, thanks for point out the process. I was worried about if all the delete tasks can be submitted by the close thread in time, before the ThreadUtils.shutdown() call its inner executor.shutdown() to let diskOperator threadpool to reject the new submitted tasks. 
So i add a join to let close thread to release cputime to diskOeprator's thread, but i think i misunderstand the fact that it is the close thread submit all delete tasks into diskOperators's pool but not the diskOperators itself do this thing. 
Now i understand that they are executed in the sequence:
 1. the close thread will submit all delete tasks into all diskOperators,
 2. then close thread will call all diskOperators.shutdown() and diskOperators.awaitTerminaion() parallel by ThreadUtils.
 3. After running at most 3000ms(default value), if timeout, all pools call diskOperators.shutdownNow().
Thus, all delete tasks must be submitted before running ThreadUtils.shutdown(), and there is no need to use join before calling ThreadUtils.shutdown() in parallel. Do I understand correctly?

Reply:
Yes, I think it's correct. To be confirmed, the 2 point can only be achieved by change
```
 diskOperators.asScala.foreach(entry => {
        ThreadUtils.shutdown(entry._2, timeout.milliseconds)
      })
```
to something like
```
ThreadUtils.parmap(diskOperators.asScala.to, "Shutdown diskOperators", diskOperators.size()) { case entry => {
    ThreadUtils.shutdown(entry._2, timeout.milliseconds)
      })
```

Reply:
Yes, will take note